src/entity/User.ts

```diff
@@ -247,101 +223,120 @@ export const updateUserEmail = async (
 const isInfoConfirmed = (user: AddUserData) =>
   !!(user.name && user.email && user.username);
 
+const handleInsertError = async (
+  error: Error,
+  req: FastifyRequest,
+  con: DataSource | EntityManager,
+  data: DeepPartial<User>,
+  maxIterations = 5,
+  iteration = 0,
+): Promise<AddNewUserResult> => {
+  req.log.error(
+    {
+      data,
+      userId: data.id,
+      error,
+      iteration,
+      maxIterations,
+    },
+    'failed to create user profile',
+  );
+  const shouldRetry = iteration < maxIterations;
+  if ('code' in error) {
+    // Unique
+    if (error.code === TypeOrmError.DUPLICATE_ENTRY) {
+      if (error.message.indexOf('users_username_unique') > -1) {
+        return {
+          status: 'failed',
+          reason: UserFailErrorKeys.UsernameEmailExists,
+        };
+      }
+
+      if (error.message.indexOf('PK_') > -1) {
+        if (shouldRetry) {
+          data.id = await generateTrackingId(req, 'user creation');
+          return safeInsertUser(req, con, data, maxIterations, iteration + 1);
+        }
+        return { status: 'failed', reason: UserFailErrorKeys.UserExists };
+      }
+
+      // If it's not username or primary key than it's twitter and github.
+      if (shouldRetry) {
+        if (error.message.indexOf('users_twitter_unique') > -1) {
+          data.twitter = null;
+        } else if (error.message.indexOf('users_github_unique') > -1) {
+          data.github = null;
+        }
+        return safeInsertUser(req, con, data, maxIterations, iteration + 1);
+      }
+    }
+  }
+
+  throw error;
+};
+
+const safeInsertUser = async (
+  req: FastifyRequest,
+  con: DataSource | EntityManager,
+  data: DeepPartial<User>,
+  maxIterations = 5,
+  iteration = 0,
+): Promise<AddNewUserResult> => {
+  try {
+    await con.getRepository(User).insert(data);
+    req.log.info(`Created profile for user with ID: ${data.id}`);
+    return { status: 'ok', userId: data.id };
+  } catch (error) {
+    return handleInsertError(error, req, con, data, maxIterations, iteration);
+  }
+};
+
 export const addNewUser = async (
   con: DataSource,
   data: AddUserData,
-  logger: FastifyBaseLogger,
+  req: FastifyRequest,
 ): Promise<AddNewUserResult> => {
   if (!checkRequiredFields(data)) {
-    logger.info({ data }, 'missing fields when adding new user');
+    req.log.info({ data }, 'missing fields when adding new user');
     return { status: 'failed', reason: UserFailErrorKeys.MissingFields };
   }
 
-  return con.transaction(async (entityManager) => {
-    const isEmailAvailable = await checkEmail(entityManager, data.email);
-    if (!isEmailAvailable) {
+  const isEmailAvailable = await checkEmail(con, data.email);
```

Comment:
Should these not run in a transaction?

Reply:
You're right I'll put it back

Reply:
Thinking about it again, there's no need for a transaction here as it doesn't protect us from anything. It will be an overkill here I guess. Also if the query fails it will rollback the transaction and will not allow me to retry